### PR TITLE
Fixed #28520 -- Added --start-at/--start-after options to runtests.py.

### PR DESCRIPTION
**Pull request by hramezani**
[ticket 28520](https://code.djangoproject.com/ticket/28520)
Continue https://github.com/django/django/pull/8957 work.

### CODE REVIEW

File: tests/runtests.py [hunk 1]
```diff
@@ -198,12 +198,22 @@ def no_available_apps(self):
...
         else:
             module_label = module_name
+
+        if not found_start:
+            if start_at and (module_label == start_at or module_label.startswith(start_at + '.')):
+                found_start = True
+            elif start_after and (module_label == start_after or module_label.startswith(start_after + '.')):
+                found_start = True
+                continue
+            elif not found_start:
```
Comment (felixxm) [inline tests/runtests.py]: Checking `found_start` looks unnecessary, maybe: 
```python
else:
     continue
```

File: tests/runtests.py [hunk 2]
```diff
@@ -198,12 +198,22 @@ def no_available_apps(self):
...
     for modpath, module_name in test_modules:
         if modpath:
             module_label = modpath + '.' + module_name
         else:
             module_label = module_name
+
+        if not found_start:
+            if start_at and (module_label == start_at or module_label.startswith(start_at + '.')):
+                found_start = True
+            elif start_after and (module_label == start_after or module_label.startswith(start_after + '.')):
```
Comment (felixxm) [inline tests/runtests.py]: We use the same logic in at least three places, I would add internal hook, e.g. `_module_match_label(self, module_label, label)`.

File: docs/internals/contributing/writing-code/unit-tests.txt [hunk 3]
```diff
@@ -213,6 +213,22 @@ Going beyond that, you can specify an individual test method like this:
 
    $ ./runtests.py --settings=path.to.settings i18n.tests.TranslationTests.test_lazy_objects
 
+Running tests from specific point
+---------------------------------
```
Comment (felixxm) [inline docs/internals/contributing/writing-code/unit-tests.txt]: I don't think that a separate section is necessary here, keeping this under `Running only some of the tests` looks good for me.

File: docs/internals/contributing/writing-code/unit-tests.txt [hunk 4]
```diff
@@ -213,6 +213,22 @@ Going beyond that, you can specify an individual test method like this:
...
+
+   $ ./runtests.py --start-at=wsgi
+
+Also, you can run the tests starting after the specified top-level module with
+``--start-after`` option:
+
+.. console::
+
+   $ ./runtests.py --start-after=wsgi
+
```
Comment (felixxm) [inline docs/internals/contributing/writing-code/unit-tests.txt]: Can we add a note about how it works with `--reverse`? i.e. that it doesn't have any impact of these options. I was confusing for me, to be honest.

Comment (hramezani) [reply docs/internals/contributing/writing-code/unit-tests.txt]: The `--reverse` option does not have any impact on these new options.
These options are for specifying the starting point of running tests. the `--reverse` option uses to reverse the order of the selected tests from the previous step.

File: tests/runtests.py [hunk 5]
```diff
@@ -495,6 +513,17 @@ def paired_tests(paired_test, options, test_labels, parallel):
...
+        print('Aborting: cannot use both --start-at and --start-after.')
+        sys.exit(1)
+    for opt_name, opt_display in {'start_at': 'start-at', 'start_after': 'start-after'}.items():
+        opt_val = getattr(options, opt_name, None)
+        if opt_val is not None:
+            if '.' in opt_val:
+                print('Aborting: --{} must be a top-level module.'.format(opt_display))
+                sys.exit(1)
+            else:
+                setattr(options, opt_name, os.path.normpath(opt_val))
```
[the remark this answers is not in the document]
Comment (felixxm) [reply tests/runtests.py]: @hramezani I think you removed `setattr(options, opt_name, os.path.normpath(opt_val))` by mistake. My proposition was to remove only `else`, i.e.
```python
if '.' in opt_val:
    print('Aborting: --%s must be a top-level module.' % opt_name.replace('_', '-'))
    sys.exit(1)
setattr(options, opt_name, os.path.normpath(opt_val))
```

[the remark this answers is not in the document]
Comment (hramezani) [reply tests/runtests.py]: Oh, Sorry, Fixed!

**Comment by felixxm** [comment]
@hramezani I split commits and made minor fixes, but I also found a use case that doesn't work properly, e.g.:
```
./runtests.py --start-after=wsgi queries
```
`queries` is not in the list of modules after `wsgi` so it will not be added to the `INSTALLED_APPS` but in the same time we pass it to the `run_tests()` via `test_labels`:
```
failures = test_runner.run_tests(
    test_labels or get_installed(),
    extra_tests=extra_tests,
)
```

**Comment by hramezani** [comment]
@felixxm Thanks for your help.
‍`queries` and any names that passed after it are considered as `test_labels` and ignores `get_installed()` in `test_labels or get_installed()`. 
We can check `test_labels` and remove items that don't exist in `get_installed()`. like this:
```
installed = get_installed()
    failures = test_runner.run_tests(
        [label for label in test_labels if label in installed] or installed,
        extra_tests=extra_tests,
    )
```

what is your idea?

**Comment by felixxm** [comment]
@hramezani I'm not sure if we need `test_labels or get_installed()` at all :thinking:  because `setup()` is already changing the list of installed apps with `test_labels`.

**Comment by hramezani** [comment]
Yes, you right I think. I think so.
do you mean to replace the `test_runner` like this?:

```
failures = test_runner.run_tests(
        get_installed(),
        extra_tests=extra_tests,
    )
```

**Comment by felixxm** [comment]
@hramezani My mistake we need `test_labels` because it contains not only module names but also paths, tests names etc.

**Comment by hramezani** [comment]
@felixxm, as you mentioned we need `test_labels`. so I think it is better to filter the first part of the `test_label`(app name) with the `installed app`.  like this:

```
installed = get_installed()
failures = test_runner.run_tests(
        [label for label in test_labels if label.split('.')[0] in installed] or installed,
        extra_tests=extra_tests,
    )
```
we have valid `installed_app` and `test_labels`

**Comment by felixxm** [comment]
@hramezani I think specifying test labels and `--start-at/--start-after` should be mutually exclusive. Because even with proposed change all modules except `queries` [will be ignored](https://github.com/django/django/blob/480492fe70b0bb7df61c00854dc8535c9d21ba64/tests/runtests.py#L210-L214).

Moreover, currently implementation doesn't work with `--bisect` and `--pair`. Please fix this.

**Comment by hramezani** [comment]
@felixxm Thanks. requested changes are done. I also added a [line](https://github.com/django/django/pull/11415/files#diff-5830044babd91b5beb690dd0104b4d56R231) to the doc about the using ``--start-at/--start-after`` and test labels together and rebased my branch.
Please let me know if anything else remains.

**Comment by felixxm** [comment]
buildbot

**Comment by felixxm** [comment]
buildbot

**Comment by felixxm** [comment]
buildbot

**Pull request merged**